### PR TITLE
feat(autoware_ptv3): handle segmentation3d decoder

### PR DESCRIPTION
**Pull request by amadeuszsz**
## Description

- Adjust naming convention (`backbone` now is `encoder`).
- Add TensorRT support for the split Segmentation3D decoder using per-stage encoder features, pooling clusters, and serialization metadata.
- Build dynamic optimization profiles using geometry-aware voxel limits for each encoder stage.
- Update Detection3D to consume the two coarsest encoder stages and validate its BEV resolution against the selected encoder stage.
- Add and validate enc_channels, dec_depths, and dec_patch_size model parameters; skip frames that cannot satisfy decoder attention requirements.
- Make TensorRT workspace sizes configurable independently for the encoder and both heads.
- Move voxels_num to runtime configuration and update package parameters, schemas, and test fixtures for the new engine interfaces.
- Expand troubleshooting guidance for TensorRT memory failures caused by workspace or voxel limits.

## Related links

**Parent Issue:**

- Link

## Notes for reviewers

- Source code keep growing, refactor PR start slowly getting necessary.
- Artifacts available in ~~[TIER IV INTERNAL LINK](https://drive.google.com/drive/folders/15bXZ2CHMQZMUUSBOW-ORq29Zg6BQVxL1?usp=sharing)~~ [TIER IV INTERNAL LINK](https://drive.google.com/drive/folders/1Gt_L3dj2RCGxeBV2GRfmrhv6LHTZMdXS?usp=sharing).
  - ~~Current link contains artifacts after pretraining stage just for debug purposes. I will update link with final artifact when fine-tuning training is finished.~~
  - Current link contains artifacts after all the 4 stages of training.
-  Voxel profiles could be tuned per stage to reduce GPU memory usage. For now, we use conservative limits because sparse pooling does not guarantee a fixed reduction in voxel count. We could optimize this later based on real data, but it can be done later, in a separate PR

## Interface changes

None.

## Effects on system behavior

None.

### CODE REVIEW

File: perception/autoware_ptv3/schema/ptv3.schema.json [hunk 1]
```diff
@@ -113,12 +143,25 @@
...
+        "workspace_size": {
+          "type": "integer",
+          "minimum": 1,
+          "description": "TensorRT builder memory-pool limit in bytes for this engine.",
+          "default": 17179869184
         }
       },
-      "required": ["use_head"],
+      "required": ["use_head", "workspace_size"],
       "allOf": [
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply perception/autoware_ptv3/schema/ptv3.schema.json]: https://github.com/autowarefoundation/autoware_universe/pull/12973/commits/88860b7e685f042af7046bce992ef366c611611c

File: perception/autoware_ptv3/schema/ptv3.schema.json [hunk 2]
```diff
@@ -150,24 +193,46 @@
...
+          "default": 4294967296
         }
       },
-      "required": ["use_head"],
+      "required": ["use_head", "workspace_size"],
       "allOf": [
         {
-          "if": { "properties": { "use_head": { "const": true } }, "required": ["use_head"] },
+          "if": {
+            "properties": {
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply perception/autoware_ptv3/schema/ptv3.schema.json]: https://github.com/autowarefoundation/autoware_universe/pull/12973/commits/88860b7e685f042af7046bce992ef366c611611c

File: perception/autoware_ptv3/lib/ptv3_trt.cpp [hunk 3]
```diff
@@ -837,19 +977,67 @@ bool PTv3TRT::preProcess(
...
+    const auto stage_count_voxels = serialized_pooling_num_voxels_[stage];
+    // The exported attention windows assume at least patch_size voxels at every block stage
+    // (training adapts the window, the baked graph does not). Skip the frame instead of
+    // silently diverging from training behavior.
+    if (stage_count_voxels < config_.dec_patch_size_[stage]) {
+      RCLCPP_ERROR_STREAM(
+        rclcpp::get_logger("ptv3"),
+        "Seg decoder stage " << stage << " has " << stage_count_voxels
+                             << " voxels, below its attention patch size "
+                             << config_.dec_patch_size_[stage] << ". Skipping frame.");
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply perception/autoware_ptv3/lib/ptv3_trt.cpp]: Nope, this is error

**Comment by mojomex** [comment]
image

Timings on RTX 4000 SFF Ada are looking good!

File: perception/autoware_ptv3/config/ptv3.param.yaml [hunk 4]
```diff
@@ -5,13 +5,16 @@
...
+    encoder:
+      onnx_path: "$(var model_path)/$(var model_variant)_encoder.onnx"
+      engine_path: "$(var model_path)/$(var model_variant)_encoder.engine"
+      workspace_size: 8589934592  # 8 GiB
+      voxels_num: [16000, 160000, 256000]
     segmentation3d:
       use_head: $(var use_seg3d_head)
       onnx_path: "$(var model_path)/$(var model_variant)_seg3d_head.onnx"
       engine_path: "$(var model_path)/$(var model_variant)_seg3d_head.engine"
+      workspace_size: 8589934592  # 8 GiB
```
Comment (mojomex) [inline perception/autoware_ptv3/config/ptv3.param.yaml:17]: Is there a strong reason why this isn't auto-computed? AFAIK the plugins should derive their scratch size from max input dims.

Having this as a config option might give users wrong hopes, and can result in failures.

File: perception/autoware_ptv3/include/autoware/ptv3/ptv3_config.hpp [hunk 5]
```diff
@@ -163,19 +193,25 @@ class PTv3Config
...
 
-      auto is_integer_multiple = [](const float numerator, const float denominator) {
-        constexpr float eps = 1e-3F;
-        const float remainder = std::fmod(numerator, denominator);
-        return std::abs(remainder) < eps || std::abs(remainder - denominator) < eps;
-      };
-
-      if (!is_integer_multiple(bbox_voxel_x_size_, voxel_x_size_)) {
+      // The exporter taps encoder stages S-2 (BEV resolution) and S-1, so the detection grid
+      // resolution must equal the voxel size at stage S-2's cumulative pooling depth.
```
Comment (mojomex) [inline perception/autoware_ptv3/include/autoware/ptv3/ptv3_config.hpp:187]: Okay to hardcode for now but these are very much tunable hyperparams, so should become part of config at some point.

**Comment by mojomex** [comment]
@codex review

File: perception/autoware_ptv3/schema/ml_package_ptv3.schema.json [hunk 6]
```diff
@@ -67,26 +77,55 @@
...
           "description": "Flat RGB palette for visualization output. Must contain 3 values per class."
+        },
+        "dec_depths": {
+          "type": "array",
+          "items": {
+            "type": "integer",
+            "minimum": 0
+          },
+          "description": "Decoder attention-block counts per stage; stages with blocks add serialization-metadata inputs to the seg-head engine.",
+          "default": [0, 0, 0, 0],
```
Comment (mojomex) [inline perception/autoware_ptv3/schema/ml_package_ptv3.schema.json:103]: Since the number of blocks does not matter for the engine interface, making this a list of bools might be clearer

File: perception/autoware_ptv3/schema/ml_package_ptv3.schema.json [hunk 7]
```diff
@@ -67,26 +77,55 @@
...
+          "minItems": 1
+        },
+        "dec_patch_size": {
+          "type": "array",
+          "items": {
+            "type": "integer",
+            "minimum": 1
+          },
+          "description": "Decoder attention window size per stage; frames with fewer voxels at a block stage are skipped.",
```
Comment (mojomex) [inline perception/autoware_ptv3/schema/ml_package_ptv3.schema.json]: The exported model seems to handle padding correctly already (see [this code in Autoware-ML](https://github.com/tier4/autoware-ml/blob/4c677b2c22e86d149b4aef49cdb93b6bbeedf0ff/autoware_ml/models/segmentation3d/backbones/ptv3.py#L412)), so we don't need `dec_patch_size` params and related skipping logic.

File: perception/autoware_ptv3/schema/ptv3.schema.json [hunk 8]
```diff
@@ -113,12 +143,25 @@
         },
         "filter": {
           "$ref": "#/definitions/filter"
+        },
+        "workspace_size": {
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply perception/autoware_ptv3/schema/ptv3.schema.json:147]: You can't do it with json schema. Also this is not important at all.

**Review by mojomex — COMMENTED** [review]
intermediate review

File: perception/autoware_ptv3/config/ptv3.param.yaml (hunk 4 above)
Comment (mojomex) [reply perception/autoware_ptv3/config/ptv3.param.yaml:17]: As discussed, this affects which tactics are available to TensorRT and is a tuning param. Closing this conversation.

File: perception/autoware_ptv3/include/autoware/ptv3/ptv3_config.hpp [hunk 9]
```diff
@@ -298,6 +334,41 @@ class PTv3Config
...
+      for (auto value = pooling_strides_[stage]; value > 1; value >>= 1) {
+        ++cumulative_depth;
+      }
+    }
+    const auto ceil_shift = [cumulative_depth](const std::int64_t size) {
+      return (size + (std::int64_t{1} << cumulative_depth) - 1) >> cumulative_depth;
+    };
+    const auto grid_cells =
+      ceil_shift(grid_x_size_) * ceil_shift(grid_y_size_) * ceil_shift(grid_z_size_);
+    return std::min(max_num_voxels_, grid_cells);
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply perception/autoware_ptv3/include/autoware/ptv3/ptv3_config.hpp:359]: https://github.com/autowarefoundation/autoware_universe/pull/12973/commits/7c75e5e0c024f2f6a236f949e9dbc99299dbac00

File: perception/autoware_ptv3/config/ptv3.param.yaml (hunk 4 above)
Comment (amadeuszsz) [reply perception/autoware_ptv3/config/ptv3.param.yaml:17]: It is about how much memory we let TensorRT use for building the engine, which might result with different strategies.

File: perception/autoware_ptv3/schema/ml_package_ptv3.schema.json (hunk 7 above)
Comment (amadeuszsz) [reply perception/autoware_ptv3/schema/ml_package_ptv3.schema.json]: Removed in https://github.com/autowarefoundation/autoware_universe/pull/12973/commits/7d7b322bae7795a1835106ec486b727dfff1741d

File: perception/autoware_ptv3/schema/ml_package_ptv3.schema.json (hunk 6 above)
Comment (amadeuszsz) [reply perception/autoware_ptv3/schema/ml_package_ptv3.schema.json:103]: I was thinking about this but it but `int` actually does not cost us anything and give you direct insight about some of model hyperparameter. If we decide to maintain two models with only different `dec_depths` e.g.: `[0, 0, 1, 1]` and `[0, 0, 2, 2]`, theoretically a separate configs does not bring anything new, but eventually we always match configs by model name so each onnx have its own `ml_package_<model_variant>_*.param.yaml` config.

File: perception/autoware_ptv3/lib/ptv3_trt.cpp [hunk 10]
```diff
@@ -371,43 +388,107 @@ void PTv3TRT::initBackboneTrt(const tensorrt_common::TrtCommonConfig & trt_confi
...
+      continue;
+    }
+    const auto prefix = "serialized_pooling_" + std::to_string(stage - 1) + "_";
+    for (const auto & field :
+         {std::string("serialized_order"), std::string("serialized_inverse")}) {
+      network_io.emplace_back(
+        prefix + field, nvinfer1::Dims{2, {num_orders, -1}}, nvinfer1::DataType::kINT64);
+      profile_dims.emplace_back(
+        prefix + field, nvinfer1::Dims{2, {num_orders, counts[0]}},
+        nvinfer1::Dims{2, {num_orders, counts[1]}}, nvinfer1::Dims{2, {num_orders, counts[2]}});
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply perception/autoware_ptv3/lib/ptv3_trt.cpp:476]: https://github.com/autowarefoundation/autoware_universe/pull/12973/commits/7d7b322bae7795a1835106ec486b727dfff1741d

File: perception/autoware_ptv3/config/ptv3.param.yaml [hunk 11]
```diff
@@ -28,17 +28,17 @@
...
-          PEDESTRIAN: [0.1, 0.1, 0.1, 0.1]
-          TRAFFIC_CONE: [0.01, 0.01, 0.01, 0.01]
-          BARRIER: [0.01, 0.01, 0.01, 0.01]
+          CAR: [0.144, 0.137, 0.123, 0.123]
+          TRUCK: [0.137, 0.138, 0.113, 0.113]
+          BUS: [0.141, 0.148, 0.125, 0.125]
+          BICYCLE: [0.178, 0.166, 0.130, 0.130]
+          PEDESTRIAN: [0.160, 0.140, 0.128, 0.128]
+          TRAFFIC_CONE: [0.169, 0.141, 0.136, 0.136]
+          BARRIER: [0.162, 0.099, 0.105, 0.105]
```
Comment (mojomex) [inline perception/autoware_ptv3/config/ptv3.param.yaml]: (nit) technically unrelated, would prefer as separate PR.

File: perception/autoware_ptv3/config/ptv3.param.yaml [hunk 12]
```diff
@@ -18,7 +18,7 @@
       source_reconstruction: "full"
       filter:
         class_probability_threshold: 0.05
-        classes: ["drivable_flat"]
+        classes: ["drivable_flat", "non_drivable_flat", "noise"]
```
Comment (mojomex) [inline perception/autoware_ptv3/config/ptv3.param.yaml]: Filtering `non_drivable_flat` is a safety hazard (vehicle will think non-drivable areas are empty).

(nit) Also, unrelated to this PR.

File: perception/autoware_ptv3/lib/preprocess/preprocess_kernel.cu [hunk 13]
```diff
@@ -214,7 +214,10 @@ __global__ void points2FeaturesKernel<CloudPointTypeXYZIRADRT>(
 
 __global__ void cropKernel(
   float4 * __restrict__ points, std::uint32_t * __restrict__ mask, int num_points, float min_x,
-  float min_y, float min_z, float max_x, float max_y, float max_z)
+  float min_y, float min_z, float max_x, float max_y, float max_z, float voxel_size_x,
+  float voxel_size_y, float voxel_size_z, std::int32_t coord_min_x, std::int32_t coord_min_y,
+  std::int32_t coord_min_z, std::int32_t grid_x_size, std::int32_t grid_y_size,
+  std::int32_t grid_z_size)
```
Comment (mojomex) [inline perception/autoware_ptv3/lib/preprocess/preprocess_kernel.cu]: This function is now not so self-explanatory anymore... logically, `{min,max}_{x,y,z}` is redundant to the voxel/coord/grid_size inputs.

Since there seem to be no other users of this function, why not revert to the `{min,max}_{x,y,z}` interface, make the `max` bounds exclusive, and set all bounds to the voxel grid boundaries?

**Review by mojomex — COMMENTED** [review]
Thanks for addressing the review comments!
Looking good except for some unrelated changes that made it in, and the now quite complex cropKernel usage.

File: perception/autoware_ptv3/config/ptv3.param.yaml (hunk 11 above)
Comment (amadeuszsz) [reply perception/autoware_ptv3/config/ptv3.param.yaml]: It is related. This PR requires update of artifacts in `awf/autoware`. These numbers are related to new artifacts.

File: perception/autoware_ptv3/config/ptv3.param.yaml (hunk 12 above)
Comment (amadeuszsz) [reply perception/autoware_ptv3/config/ptv3.param.yaml]: Agree, wanted to smuggle this tiny change 🤣 . Reverted in https://github.com/autowarefoundation/autoware_universe/pull/12973/commits/db62a4fe413247e23de9c8c2cad50dd1b11713af

> Filtering `non_drivable_flat` is a safety hazard (vehicle will think non-drivable areas are empty).

It replicates typical ground segmentation behavior, so I believe we are good with filtering it out. Additionally, model sometimes is confused with both flat-like classes so we need to filter both classes. Anyway, I will add it in a separate PR.

File: perception/autoware_ptv3/lib/preprocess/preprocess_kernel.cu (hunk 13 above)
Comment (amadeuszsz) [reply perception/autoware_ptv3/lib/preprocess/preprocess_kernel.cu]: Reverted in https://github.com/autowarefoundation/autoware_universe/pull/12973/commits/e71b9034094682b5beb014aa6bb84c353a2cb8ad

Comment (amadeuszsz) [reply perception/autoware_ptv3/lib/preprocess/preprocess_kernel.cu]: Note: Current crop assumes `max == (coord_min + grid_size) * voxel_size`. If `round((max - min) / voxel)` rounds **down**, `max` exceeds the grid edge and points there voxelize to `coord == grid_size` (out of grid), with no guard to reject them since `coord < grid_size` is not used.

Clean fix: clamp crop bounds to the grid edge, or assert grid alignment in `PTv3Config`.

Since this is not related to the scope of this change, we can address a separate PR.

File: perception/autoware_ptv3/config/ptv3.param.yaml (hunk 11 above)
Comment (amadeuszsz) [reply perception/autoware_ptv3/config/ptv3.param.yaml]: Buuuut, eventually I changed my mind. I do not want to update this configs each time I upload a new artifacts. Let's keep simply `0.1`. Correct threshold we will handle internally.
https://github.com/autowarefoundation/autoware_universe/pull/12973/commits/98838784d644bca486846bbc752f2eaf7cb98e87

**Comment by mojomex** [comment]
Will test today, sorry for the wait!

**Review by mojomex — APPROVED** [review]
LGTM!

https://github.com/user-attachments/assets/418bb376-844a-44d3-92af-727b8b02915b
